Improve usability of Directory datatype

wm75 opened the pull request:
<del>This adds functionality to the Directory datatype class, which can now be displayed and downloaded as an archive.</del>
It also adds a new `archive_to_directory` converter that generalizes the existing tar_to_directory one to work with tar and zip archives. Also updates the older converter's requirement to an existing version of the galaxy-util package. Previously the exact requirement wasn't installable via conda.

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [x] Instructions for manual testing are as follows:
  1. upload an archive as either zip or tar
  2. convert the datatype to directory
  3. explore

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

wm75 committed:
Improve usability of Directory datatype

bernt-matthias commented on lib/galaxy/datatypes/converters/archive_to_directory.xml:
```diff
@@ -0,0 +1,33 @@
+}]]></configfile>
+    </configfiles>
+    <inputs>
+        <param format="tar,zip" name="input1" type="data"/>
```
Should we add `tar.gz` or `tar.bz2`?

Edit: seems that we do not have these datatypes yet, or?

bernt-matthias commented on lib/galaxy/config/sample/datatypes_conf.xml.sample:
```diff
@@ -278,13 +278,14 @@
+        <converter file="archive_to_directory.xml" target_datatype="directory" />
+    </datatype>
     <datatype extension="ncbi_genome_dataset.zip" type="galaxy.datatypes.binary:CompressedZipArchive" subclass="true" display_in_upload="true"/>
     <datatype extension="tar" type="galaxy.datatypes.binary:CompressedArchive" subclass="true" display_in_upload="true">
```
Wondering if this could be `auto_compressed_types="gz,bz2"`?

wm75 replied:
for "tar" you mean, to decompress compressed archives on upload?

wm75 replied:
added it and seems to work, thanks

bernt-matthias commented on lib/galaxy/datatypes/data.py:
```diff
@@ -1205,6 +1205,143 @@ def regex_line_dataprovider(
+        for dirpath, dirnames, filenames in os.walk(root_path):
+            for f in filenames:
+                fp = os.path.join(dirpath, f)
+                # skip if it is symbolic link
```
Should we allow links in directories at all?

wm75 replied:
So, for directories created via the converter, links are checked for safety, and excluded if they point to higher up than the top-level folder of their archive.
If we had tools generating directories, then I don't think we currently have a protection mechanism, but would have to investigate the tool output manually?

mvdbeek replied:
I don't think that's the responsibility of a datatype, and I'm pretty sure that can't be moved into the object store, and if it can that would be the right place to check this.

bernt-matthias commented on lib/galaxy/datatypes/data.py:
```diff
@@ -1205,6 +1205,143 @@ def regex_line_dataprovider(
 class Directory(Data):
```
Why does the `Directory` type does not define an `ext` attribute?

Do you know how the data is stored on disk? Is it zipped? Since one application of this is to hold large amounts of files it would actually be cool .. I guess. 

Where in the code is this managed?

wm75 replied:
> Why does the Directory type does not define an ext attribute?

Oh, I guess it *should* define one. There just wasn't one initially and I didn't even think about it.



> Do you know how the data is stored on disk? Is it zipped? Since one application of this is to hold large amounts of files it would actually be cool .. I guess.
> 
> Where in the code is this managed?

The boring answer is: there's nothing of that type of magic going on here. The whole folder structure is simply stored at the extra_files_path. Some clever compression might be nice to have, yes, but I have no clue how you would implement this efficiently without knowing how often files get used, which depends on what tools you run on that data and what the user wants to do with it (viewing, downloading, etc.). So this sounds like something for a *really* clever datatype class, not like something I would implement.

mvdbeek replied:
Compression is not galaxy's responsibility, let's not go down that route.

wm75 replied:
added the file_ext attribute

bernt-matthias reviewed (commented):
It would be great to have some unit tests for setting metadata. Maybe here https://github.com/galaxyproject/galaxy/blob/dev/test/unit/data/datatypes/test_data.py. 

There should be a bit of inspiration for such tests in the folder.

mvdbeek commented on lib/galaxy/datatypes/data.py:
```diff
@@ -1205,6 +1205,143 @@ def regex_line_dataprovider(
 class Directory(Data):
     """Class representing a directory of files."""
 
+    # The behavior of this class is intended to be similar to that of
```
It should really just be a directory, can you make all that fancy stuff a subclass and include some specific implementation ? We're making heavy use of this in the CWL branch. See also https://github.com/galaxyproject/galaxy/compare/dev...mvdbeek:galaxy:directory_datatype_improvements for details.

wm75 replied:
well, what exactly do you mean? It really *is* just a directory, but one with minimal knowledge of displaying its contents and size? Why would that be a problem for the CWL branch?

But sure, I can turn it into a SmartDirectory subclass or something if you prefer, I would just like to understand why that's an advantage.

wm75 replied:
I could also just add your:

https://github.com/mvdbeek/galaxy/blob/fc7e959e8fc1a271131759ea26fc18a5f3ced6ad/lib/galaxy/datatypes/data.py#L1214-L1219

and then there should be no change in behavior for you.

I even had written your
https://github.com/mvdbeek/galaxy/blob/fc7e959e8fc1a271131759ea26fc18a5f3ced6ad/lib/galaxy/datatypes/data.py#L467

at first, but that won't work with subclasses. That's why I went the somewhat more laborious route via display_data.

mvdbeek replied:
>  but that won't work with subclasse

There's surely a better way than repeating the code.

> Why would that be a problem for the CWL branch?

It's not a problem for the CWL branch, but I don't think we should store that in the database. Literal filesystem directories don't have this, and the total file size is redundant with what we calculate and store anyway. If this is important for a specific datatype and the tools consuming this then they can implement that.

wm75 replied:
regarding the file size: isn't that something the user would want to know?

mvdbeek replied:
And they can, as for any other file.

wm75 replied:
how?

mvdbeek replied:
as for any other dataset, it's a property of the dataset, https://github.com/mvdbeek/galaxy/blob/4048dc9a332e8c051caebb79ac441402f9ebf693/lib/galaxy/managers/datasets.py#L261

wm75 replied:
Aha, so you're saying the class can make use of that property instead of calculating it itself? That'd be nice for sure!

wm75 replied:
It works. Thanks for the hint!

wm75 commented on lib/galaxy/datatypes/data.py:
```diff
@@ -1205,6 +1205,143 @@ def regex_line_dataprovider(
+        if to_ext:
+            # Download the directory structure as an archive
+            trans.log_event(f"Download directory for dataset id: {str(dataset.id)}")
+            return self._archive_composite_dataset(trans, dataset, headers, do_action=kwd.get("do_action", "zip"))
```
@mvdbeek this is the code that gets repeated to manage the downloading as an archive :)

The rest is for making the eye icon work.

mvdbeek replied:
Sorry, not following, what do you mean by making the eye icon work ?

mvdbeek replied:
The index stuff doesn't belong here, directories don't have indexes.

wm75 replied:
yeah, that's the part that will be useful in subclasses (like zarr). That's why the candidate index files list for the parent class is empty. Could be implemented in just relevant subclasses of course.

wm75 replied:
> Sorry, not following, what do you mean by making the eye icon work ?

well, displaying the root folder or the index file. With the current version of directory, when you click the eye icon, you'll just an empty file downloaded.

mvdbeek replied:
> Could be implemented in just relevant subclasses of course.

But a directory is a relevant, self-contained, out-there-in-the-wild datatype. Let's not add data where none is needed. subclasses should implement this if required.

mvdbeek replied:
> With the current version of directory, when you click the eye icon, you'll just an empty file downloaded.

Feel free to disable the eye icon on the client or implement a directory browser, but this does not belong into the datatype.

mvdbeek commented on lib/galaxy/datatypes/data.py:
```diff
@@ -1205,6 +1205,143 @@ def regex_line_dataprovider(
+            index_file = dataset.metadata.index_file
+            if not filename or filename == 'index':
+                if root_folder is not None and index_file:
+                    # display the index file in lieu of the empty primary dataset
```
That's not the right layer to do this, that's a client concern.

wm75 committed:
Use dataset's total_size property instead of calculating it again

wm75 committed:
Add auto-converters for tar.gz and tar.bz2

as suggested by @bernt-matthias

wm75 committed:
Make new converter testable

wm75 committed:
Add file_ext and missing type annotation

wm75 committed:
Fix linting

wm75 committed:
Fix code formatting

davelopez committed:
Drop index-related metadata from Directory datatype

This will go into a sub-datatype when needed.

mvdbeek commented on lib/galaxy/datatypes/data.py:
```diff
@@ -1212,6 +1212,51 @@ def regex_line_dataprovider(
+    # root folder inside the directory. The root_folder attribute lets tools
+    # access that folder without first exploring the directory tree themselves.
+    # Will be set to the empty string for flat directories.
+    MetadataElement(
```
Can we remove this ? The root directory should just be `$input.extra_files_path`

davelopez replied:
Sure. I'm trying to remember why this was needed. I think it was because some zip files might compress the "target directory" or a "container directory" and we wanted to get the "target directory" in any case, but I see this should not have anything to do with the "base" Directory datatype and probably it should go into a sub-class when needed :+1:

davelopez committed:
Drop root_folder metadata from Directory datatype

This should probably go in sub-classes that expect specific directory structures.

mvdbeek commented on lib/galaxy/datatypes/data.py:
```diff
@@ -1217,31 +1217,9 @@ class Directory(Data):
     def set_peek(self, dataset: DatasetProtocol, **kwd) -> None:
         if not dataset.dataset.purged:
-            dataset.peek = f"{dataset.metadata.root_folder}"
+            dataset.peek = f"{dataset.extra_files_path}"
```
That's the file system path. No peek seems fine here ?

davelopez committed:
Do not set peek for directories

davelopez committed:
Set output format to directory

davelopez commented on lib/galaxy/datatypes/converters/archive_to_directory.xml:
```diff
@@ -0,0 +1,37 @@
+        python -c "from galaxy.util.compression_utils import CompressedFile; CompressedFile('$input1').extract('.')"
+    ]]></command>
+    <configfiles>
+        <configfile filename="metadata_json"><![CDATA[{
+    "output1": {
+        "name": "$input1.name unpacked to $__target_datatype__",
+        "ext": "$__target_datatype__"
+    }
+}]]></configfile>
```
Hmm, somehow this doesn't work anymore...

When I run the converter to `directory` the datatype changes to `directory`, but I don't get the updated name or the blurb with the total size set in peek. I'm not sure what I'm missing here... :thinking: 

![image](https://github.com/user-attachments/assets/a47a2b2b-2b52-4b0b-afaa-f57963776447)

davelopez replied:
It looks like I was missing `cp '$metadata_json' galaxy.json` in the command, but now I get a template error:

```
Traceback (most recent call last):
  File "/home/dlopez/dev/galaxy/lib/galaxy/tools/evaluation.py", line 107, in global_tool_logs
    return func()
           ^^^^^^
  File "/home/dlopez/dev/galaxy/lib/galaxy/tools/evaluation.py", line 634, in _build_command_line
    command_line = fill_template(
                   ^^^^^^^^^^^^^^
  File "/home/dlopez/dev/galaxy/lib/galaxy/util/template.py", line 115, in fill_template
    return fill_template(
           ^^^^^^^^^^^^^^
  File "/home/dlopez/dev/galaxy/lib/galaxy/util/template.py", line 115, in fill_template
    return fill_template(
           ^^^^^^^^^^^^^^
  File "/home/dlopez/dev/galaxy/lib/galaxy/util/template.py", line 115, in fill_template
    return fill_template(
           ^^^^^^^^^^^^^^
  [Previous line repeated 7 more times]
  File "/home/dlopez/dev/galaxy/lib/galaxy/util/template.py", line 123, in fill_template
    raise first_exception or e
  File "/home/dlopez/dev/galaxy/lib/galaxy/util/template.py", line 87, in fill_template
    return unicodify(t, log_exception=False)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/dlopez/dev/galaxy/lib/galaxy/util/__init__.py", line 1182, in unicodify
    value = str(value)
            ^^^^^^^^^^
  File "/home/dlopez/dev/galaxy/.venv/lib/python3.11/site-packages/Cheetah/Template.py", line 1053, in __unicode__
    return getattr(self, mainMethName)()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "cheetah_DynamicallyCompiledCheetahTemplate_1728642092_1426992_84394.py", line 87, in respond
NameMapper.NotFound: cannot find 'metadata_json'
```

mvdbeek replied:
It's just `cp metadata_json galaxy.json` if you're using the `filename` attribute, the previous version would only work with `<configfile name="metadata_json">`

davelopez replied:
Thank you! That fixed the template error, but the metadata is still not being processed for some reason...

There is also an error when trying to cleanup the job directory:

```
galaxy.jobs ERROR 2024-10-11 12:49:34,434 [pN:main.1,p:229474,tN:LocalRunner.work_thread-0] Unable to cleanup job 11652
Traceback (most recent call last):
  File "/home/dlopez/dev/galaxy/lib/galaxy/jobs/__init__.py", line 2160, in cleanup
    self.object_store.delete(self.get_job(), base_dir="job_work", entire_dir=True, obj_dir=True)
  File "/home/dlopez/dev/galaxy/lib/galaxy/objectstore/__init__.py", line 580, in delete
    return self._invoke(
           ^^^^^^^^^^^^^
  File "/home/dlopez/dev/galaxy/lib/galaxy/objectstore/__init__.py", line 485, in _invoke
    return self.__getattribute__(f"_{delegate}")(obj=obj, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/dlopez/dev/galaxy/lib/galaxy/objectstore/__init__.py", line 1202, in _delete
    return self._call_method("_delete", obj, False, False, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/dlopez/dev/galaxy/lib/galaxy/objectstore/__init__.py", line 1524, in _call_method
    return self._resolve_backend(object_store_id).__getattribute__(method)(obj, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/dlopez/dev/galaxy/lib/galaxy/objectstore/__init__.py", line 1068, in _delete
    path = self._get_filename(obj, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/dlopez/dev/galaxy/lib/galaxy/objectstore/__init__.py", line 1114, in _get_filename
    raise ObjectNotFound
galaxy.exceptions.ObjectNotFound: No such object found.
```

I guess it is expecting a "main" file for the dataset?

mvdbeek replied:
That's an attempt to delete the job working directory, I don't think this is related.

mvdbeek replied:
`dataset.blurb = "done"` ... this is where the done comes from. I would check out then why the datatype-specific metadata isn't being set. That said ... why use the galaxy.json method at all if you know the output format up front ?

davelopez replied:
Yeah sorry I wasn't clear. The metadata not being set is, of course, not related at all to this other error attempting to cleanup the job directory. I thought it might have something to do with the `directory` datatype itself and worth mentioning.

Going back to the issue with the converter output metadata:

I thought that using `<configfile filename="metadata_json">` and then `<outputs provided_metadata_file="metadata_json">` will directly work (without having to `cp` any file). But I must have misunderstood something because it didn't work.

Curiously enough, the only way I got the metadata to work is by trying to `cp metadata_json metadata_json`, which doesn't make sense to me and obviously gives a command error but is the only way I got the output name changed...

![image](https://github.com/user-attachments/assets/851bb8f1-6414-4715-8104-b2763bf75537)

So definitely I'm missing something silly.

davelopez replied:
> why use the galaxy.json method at all if you know the output format up front ?

That is a good question. I'm trying to continue what was already done so I don't know the exact reason. I guess it is to set the output name to "X unpacked to Y", and the target format. I assume right now the format will be `directory` but when we have the `zarr` datatype it should be able to set it to that as well so maybe we can't assume it is always `directory`.

Is there a better way?

mvdbeek replied:
You can just use a normal label attribute on the output to do that.

> I thought that using `<configfile filename="metadata_json">` and then `<outputs provided_metadata_file="metadata_json">` will directly work (without having to `cp` any file). But I must have misunderstood something because it didn't work.

That is the idea, see also https://github.com/galaxyproject/galaxy/blob/dev/test/functional/tools/tool_provided_metadata_9.xml#L20-L22

davelopez replied:
> That is the idea, see also https://github.com/galaxyproject/galaxy/blob/dev/test/functional/tools/tool_provided_metadata_9.xml#L20-L22

I saw that example, and I did "exactly" the same thing (even renamed variables to match it), but unfortunately, it didn't work, so there must be something slightly different going on... I'll keep investigating...

davelopez replied:
Turns out the converter is perfectly fine, the issue was caused by setting `dataset.blurb = nice_size(dataset.dataset.total_size)`  in https://github.com/galaxyproject/galaxy/pull/17614/files#diff-38a85eadce6179cb91c77072b0617e6a8f652869c5b3fb1d76740909c3f24027R1222 (as you were hinting) if I replace the blurb with something else like "Directory of files" everything works... 

![image](https://github.com/user-attachments/assets/a4f205b8-b33c-4711-9137-558b199e1210)


So I understand this is not the correct way of getting the total uncompressed size?

mvdbeek replied:
I mean this is kind of a useless blurb considering we store the total size elsewhere. There's a try/except that might hide the real exception.

davelopez replied:
I see... the "done" blurb was extremely confusing, indeed it shows when there is an exception in the set_peek function, I'm not sure if that is a bug or by design.

If this blurb is not useful, then I will just drop the entire `set_peek`. Thanks!

davelopez committed:
Copy metadata to galaxy.json in archive_to_directory converter

mvdbeek commented on lib/galaxy/datatypes/converters/archive_to_directory.xml:
```diff
@@ -22,8 +23,8 @@
-    <outputs provided_metadata_file="metadata_json">
-        <data format="directory" name="output1"/>
+    <outputs>
+        <data format="auto" name="output1"/>
```
The previous version is a lot easier to reason about in the workflow editor and when queuing up jobs, what was wrong with it ?

davelopez replied:
I was trying to make it work by looking at other converters, see https://github.com/galaxyproject/galaxy/pull/17614#discussion_r1796639347

But that didn't help... so I can revert back. Nothing I tried so far worked for some reason I fail to see... :disappointed:

davelopez committed:
Revert back to use provided_metadata_file

davelopez committed:
Remove set_peek method from Directory datatype

mvdbeek commented on lib/galaxy/datatypes/data.py:
```diff
@@ -1212,6 +1212,18 @@ def regex_line_dataprovider(
 
+    file_ext = "directory"
+
+    def _archive_main_file(
```
Can you add a test for roundtripping a directory via the API ?

davelopez replied:
Sure I can try!

davelopez replied:
I've added a test, but I'm not sure why I'm getting `ModuleNotFoundError: No module named 'galaxy'` when running the converter tool... using the UI seems to work fine :thinking:

mvdbeek replied:
I don't think the API tests are setting up dependency resolution. I am mostly interested in verifying that the structure of the tar archive is the same pre-and-post upload. In fact I think even the checksum should match if we're not compressing the archive. In either case if you upload a tar file and download it again that should be sufficient. The converter is tested in the test framework.

davelopez replied:
Ahh I see, so you mean something like this https://github.com/galaxyproject/galaxy/commit/fc7e959e8fc1a271131759ea26fc18a5f3ced6ad#diff-a6ab1700bcef9e1585a2bb0f84e8888470a770fb81c3e0337930e7cad573093fR662

I'll do that :+1:

davelopez replied:
I've tried this:

```python
    def test_fetch_directory(self, history_id):
        testdir = TestDataResolver().get_filename("testdir.tar")
        with open(testdir, "rb") as fh:
            details = self._upload_and_get_details(
                fh, api="fetch", history_id=history_id, ext="directory", assert_ok=True
            )
        assert details["file_ext"] == "directory"
        assert details["file_size"] == 3584
        content = self.dataset_populator.get_history_dataset_content(
            history_id, dataset=details, to_ext="directory", type="bytes"
        )
        dir_path = decompress_bytes_to_directory(cast(bytes, content))
        assert dir_path.endswith("testdir")
        for path, entry_class in EXPECTED_CONTENTS.items():
            path = os.path.join(dir_path, os.path.pardir, path)
            if entry_class == "Directory":
                assert os.path.isdir(path)
            else:
                assert os.path.isfile(path)
```

But if I don't run the converter manually instead of the `to_ext="directory"` the extra_files_path is empty, I guess that is why you have more changes persisting extra files in the object store in your referenced branch https://github.com/galaxyproject/galaxy/compare/dev...mvdbeek:galaxy:directory_datatype_improvements#diff-8640d91ef47bca302b00039012979f4b1b79f5dbffbe2431bc9a05f19fb4c7d0R132

Should we merge your branch instead? Is something still missing in your branch or should that be how to do it?

Sorry, I'm a bit lost :sweat_smile:

davelopez replied:
@mvdbeek, re-reading your comment:

> In either case if you upload a tar file and download it again that should be sufficient.

 do you mean something simpler like this instead?

```python
def test_upload_tar_roundtrip(self, history_id):
        testdir = TestDataResolver().get_filename("testdir.tar")
        expected_hash = md5_hash_file(testdir)
        expected_size = os.path.getsize(testdir)
        with open(testdir, "rb") as fh:
            details = self._upload_and_get_details(fh, api="fetch", history_id=history_id, assert_ok=True)
        assert details["file_ext"] == "tar"
        assert details["file_size"] == expected_size
        content = cast(
            bytes, self.dataset_populator.get_history_dataset_content(history_id, dataset=details, type="bytes")
        )
        assert len(content) == expected_size
        dir_path = decompress_bytes_to_directory(content)
        expected_contents = {
            "testdir": "Directory",
            "testdir/c": "Directory",
            "testdir/a": "File",
            "testdir/b": "File",
            "testdir/c/d": "File",
        }
        assert dir_path.endswith("testdir")
        for path, entry_class in expected_contents.items():
            path = os.path.join(dir_path, os.path.pardir, path)
            if entry_class == "Directory":
                assert os.path.isdir(path)
            else:
                assert os.path.isfile(path)

        with tempfile.NamedTemporaryFile("wb") as temp:
            temp.write(content)
            actual_hash = md5_hash_file(temp.name)
            assert actual_hash == expected_hash
```

If this is what you mean, the uploaded vs downloaded tar size and contents match, but the hashes don't (not sure why). 
I still don't see the connection with the `directory` datatype or the converter changes in this PR, so I might be misunderstanding something :disappointed:

mvdbeek replied:
You've overwritten `_archive_main_file`, as a result you've made sure you're not getting an (empty) extra file added to the archive.

`test_upload_tar_roundtrip` looks fine to me, you sure you don't need to flush and that that is why the checksums don't match ? You could also skip the hash and just compare the bytes. If the contents are the same, no added or removed files or structure then it's all good.

davelopez replied:
Ahhh that was it... I was missing the flush :see_no_evil: 

Thank you very much!!

davelopez committed:
Add roundtrip test for downloading directories

Compressed (Upload) -> Directory (Unpack) -> Compressed (Download)

davelopez committed:
Add TAR and ZIP roundtrip tests

davelopez committed:
Flush temporary file before testing hashes
